syscall-restrict/syscall-restrict.c

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2022 The CapableVMs "CHERI Examples" Contributors.
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#include <cheriintrin.h>
+
+#include <cheri/cherireg.h>
+
+#define STR(s) #s
+#define STR_VALUE(s) STR(s)
+
+typedef int (*stat_fn)(const char *restrict path, struct stat *restrict statbuf);
+
+void try_stat(stat_fn fn, char const *file)
+{
+	struct stat statbuf;
+	printf("    stat(\"%s\", &statbuf)\n", file);
+	if (fn(file, &statbuf) == 0)
+	{
+		// Print some easily-verifiable information to demonstrate that
+		// the syscall actually worked.
+		printf("    - Success, size is %zu bytes.\n", statbuf.st_size);
+	}
+	else
+	{
+		printf("    - Failed, errno = %d: %s\n", errno, strerror(errno));
+	}
+}
+
+int stat_without_perm_syscall(const char *restrict path, struct stat *restrict statbuf);
+#if __has_feature(capabilities) && defined(__aarch64__)
+#ifdef __CHERI_PURE_CAPABILITY__
+// This probably requires a static build, or at least non-sealed access to the
+// underlying 'stat'. Even then, whilst it demonstrates CHERI_PERM_SYSCALL, its
+// security is incomplete because the sentry (with the permission) is present in
+// every compartment's GOT.
+#error Unimplemented: purecap support.
+#else
+// clang-format off
+asm(	".type stat_without_perm_syscall, @function\n"
+	"stat_without_perm_syscall:\n"
+	// x0: path (passed through unmodified)
+	// x1: statbuf (passed through unmodified)
+	// First, preserve the current PCC. This requires C64.
+	"	bx	#4\n"
+	"	adr	x16, stat\n"  // In C64, this is `adr c16, stat`.
+	// Note that c16 actually refers to the PLT, not to `stat()` itself.
+	// That doesn't matter for hybrid mode, since only the PCC address is
+	// changed, but it makes this difficult to implement for purecap.
+	"	bx	#4\n"
+	"	stp	clr, c16, [sp, #-32]!\n"
+	// Now drop the permission and call 'stat'. Note that this is not
+	// secure; a malicious callee could retrieve the old PCC from the stack.
+	// However, it's sufficient for a demonstration.
+	"	mov	x17, #" STR_VALUE(CHERI_PERM_SYSCALL) "\n"
+	"	clrperm c16, c16, x17\n"
+	"	blr	c16\n"
+	// Restore and return.
+	"	ldp	clr, c16, [sp], #32\n"
+	"	gcvalue	lr, clr\n"
+	"	scvalue	clr, c16, lr\n"
```

Comment:
Oh, right. This is hybrid, so on entry `clr` has zero metadata, and just a 64-bit address. I preserve and restore the whole thing, rather than just `lr`, because it saves a coupe of instructions. To return, I'm building a full capability in `clr`, using the original address and the saved PCC metadata (in `c16`).

I think this could be made clearer by storing and restoring only `lr`, not `clr`. Clarity is more important here than the number of instructions, I think.

The `gcvalue` isn't strictly needed, and especially doesn't make sense if I make the change I just suggested.